File: WeatherAPI.pas
unit WeatherAPI;

interface

uses SysUtils, Classes, fphttpclient;

type
  TForecast = record
    Date: string;
    DateLabel: string;
    Telop: string;
    Weather: string;
  end;

  TResult = record
    Title: string;
    Forecasts: array of TForecast;
  end;

function Fetch(const City: string): TResult;

implementation

function Extract(const S, Key: string): string;
var i, j: Integer;
begin
  i := Pos(Key, S);
  if i = 0 then Exit('');
  Inc(i, Length(Key));
  j := PosEx('"', S, i);
  Result := Copy(S, i, j - i);
end;

function ExtractFrom(const S, Key: string; Start: Integer): string;
var i, j: Integer;
begin
  i := PosEx(Key, S, Start);
  if i = 0 then Exit('');
  Inc(i, Length(Key));
  j := PosEx('"', S, i);
  Result := Copy(S, i, j - i);
end;

function Parse(const Json: string): TResult;
var posi: Integer;
    f: TForecast;
begin
  Result.Title := Extract(Json, '"title":"');

  posi := 1;
  while True do
  begin
    posi := PosEx('"date":"', Json, posi);
    if posi = 0 then Break;

    f.Date := ExtractFrom(Json, '"date":"', posi);
    f.DateLabel := ExtractFrom(Json, '"dateLabel":"', posi);
    f.Telop := ExtractFrom(Json, '"telop":"', posi);
    f.Weather := ExtractFrom(Json, '"weather":"', posi);

    SetLength(Result.Forecasts, Length(Result.Forecasts) + 1);
    Result.Forecasts[High(Result.Forecasts)] := f;

    Inc(posi, 10);
  end;
end;

function Fetch(const City: string): TResult;
var http: TFPHTTPClient;
    data: string;
begin
  http := TFPHTTPClient.Create(nil);
  try
    http.Timeout := 10000;
    data := http.Get('https://weather.tsukumijima.net/api/forecast?city=' + City);
    Result := Parse(data);
  finally
    http.Free;
  end;
end;

end.

File: WeatherAPI.java
package wapi;

import java.net.URI;
import java.net.http.*;
import java.util.*;

public class WeatherAPI {

    // ===== 構造 =====
    public static class Temperature {
        public String celsius, fahrenheit;
    }

    public static class Forecast {
        public String date, dateLabel, telop;
        public String weather;
    }

    public static class Result {
        public String title;
        public List<Forecast> forecasts = new ArrayList<>();
    }

    // ===== API =====
    public static Result fetch(String city) throws Exception {
        var client = HttpClient.newHttpClient();
        var req = HttpRequest.newBuilder()
                .uri(URI.create("https://weather.tsukumijima.net/api/forecast?city=" + city))
                .GET().build();

        var res = client.send(req, HttpResponse.BodyHandlers.ofString());
        return parse(res.body());
    }

    // ===== 超軽量JSONパース =====
    private static Result parse(String json) {
        Result r = new Result();

        r.title = extract(json, "\"title\":\"");

        int pos = 0;
        while ((pos = json.indexOf("\"date\":\"", pos)) != -1) {
            Forecast f = new Forecast();

            f.date = extractFrom(json, "\"date\":\"", pos);
            f.dateLabel = extractFrom(json, "\"dateLabel\":\"", pos);
            f.telop = extractFrom(json, "\"telop\":\"", pos);
            f.weather = extractFrom(json, "\"weather\":\"", pos);

            r.forecasts.add(f);
            pos += 10;
        }

        return r;
    }

    private static String extract(String s, String key) {
        int i = s.indexOf(key);
        if (i < 0) return "";
        i += key.length();
        int j = s.indexOf('"', i);
        return s.substring(i, j);
    }

    private static String extractFrom(String s, String key, int from) {
        int i = s.indexOf(key, from);
        if (i < 0) return "";
        i += key.length();
        int j = s.indexOf('"', i);
        return s.substring(i, j);
    }
}